// snyk/dist/lib/plugins/yarn/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var snyk = require('../../');
var fileSystem = require('fs');
var fs = require('then-fs');
var path = require('path');
var lockFileParser = require('snyk-nodejs-lockfile-parser');
var debug = require('debug')('snyk');
module.exports = {
    inspect: inspect,
};
function inspect(root, targetFile, options) {
    return __awaiter(this, void 0, void 0, function () {
        var isLockFileBased, isShrinkwrapPresent, isNodeModulesFolderPresent;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    isLockFileBased = targetFile.endsWith('yarn.lock');
                    isShrinkwrapPresent = fileSystem.existsSync(path.resolve(root, targetFile).dir, 'npm-shrinkwrap.json');
                    if (!(getRuntimeVersion() < 6)) return [3 /*break*/, 2];
                    options.traverseNodeModules = true;
                    debug('Falling back to node_modules traversal for yarn.lock projects on Node < 6');
                    return [4 /*yield*/, fs.exists(path.join(root, 'node_modules'))];
                case 1:
                    isNodeModulesFolderPresent = _a.sent();
                    if (!isNodeModulesFolderPresent) {
                        // throw a custom error
                        throw new Error('Missing node_modules folder: we can\'t test ' +
                            'without dependencies.\nPlease run \'yarn install\' first.');
                    }
                    _a.label = 2;
                case 2:
                    if (isLockFileBased && !isShrinkwrapPresent && !options.traverseNodeModules) {
                        return [2 /*return*/, generateDependenciesFromLockfile(root, options, targetFile)
                                .then(function (modules) { return ({
                                plugin: {
                                    name: 'snyk-nodejs-lockfile-parser',
                                    runtime: process.version,
                                },
                                package: modules,
                            }); })];
                    }
                    // traverse node modules
                    return [2 /*return*/, snyk.modules(root, Object.assign({}, options, { noFromArrays: true }))
                            .then(function (modules) {
                            return {
                                plugin: {
                                    name: 'snyk-resolve-deps',
                                    runtime: process.version,
                                },
                                package: modules,
                            };
                        })];
            }
        });
    });
}
function generateDependenciesFromLockfile(root, options, targetFile) {
    return __awaiter(this, void 0, void 0, function () {
        var lockFileFullPath, fullPath, manifestFileFullPath, manifestFile, lockFile;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    lockFileFullPath = path.resolve(root, targetFile);
                    if (!fileSystem.existsSync(lockFileFullPath)) {
                        throw new Error('Lockfile ' + targetFile + ' not found at location: ' +
                            lockFileFullPath);
                    }
                    fullPath = path.parse(lockFileFullPath);
                    manifestFileFullPath = path.resolve(fullPath.dir, 'package.json');
                    if (!fileSystem.existsSync(manifestFileFullPath)) {
                        throw new Error('Manifest file yarn.lock not found at location: ' +
                            manifestFileFullPath);
                    }
                    if (!manifestFileFullPath && lockFileFullPath) {
                        throw new Error('Detected a lockfile at location: '
                            + lockFileFullPath + '\n However the yarn.lock is missing!');
                    }
                    return [4 /*yield*/, fs.readFile(manifestFileFullPath, 'utf-8')];
                case 1:
                    manifestFile = _a.sent();
                    return [4 /*yield*/, fs.readFile(lockFileFullPath, 'utf-8')];
                case 2:
                    lockFile = _a.sent();
                    return [2 /*return*/, lockFileParser
                            .buildDepTree(manifestFile, lockFile, options.dev, lockFileParser.LockfileType.yarn)];
            }
        });
    });
}
function getRuntimeVersion() {
    return parseInt(process.version.slice(1).split('.')[0], 10);
}
