// snyk-docker-plugin/dist/analyzer/apk-analyzer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var docker_1 = require("../docker");
function analyze(targetImage) {
    return getPackages(targetImage)
        .then(function (pkgs) { return ({
        Image: targetImage,
        AnalyzeType: 'Apk',
        Analysis: pkgs,
    }); });
}
exports.analyze = analyze;
function getPackages(targetImage) {
    return new docker_1.Docker(targetImage)
        .catSafe('/lib/apk/db/installed')
        .then(parseFile);
}
function parseFile(text) {
    var pkgs = [];
    var curPkg = null;
    for (var _i = 0, _a = text.split('\n'); _i < _a.length; _i++) {
        var line = _a[_i];
        curPkg = parseLine(line, curPkg, pkgs);
    }
    return pkgs;
}
function parseLine(text, curPkg, pkgs) {
    var key = text.charAt(0);
    var value = text.substr(2);
    switch (key) {
        case 'P': // Package
            curPkg = {
                Name: value,
                Version: undefined,
                Source: undefined,
                Provides: [],
                Deps: {},
                AutoInstalled: undefined,
            };
            pkgs.push(curPkg);
            break;
        case 'V': // Version
            curPkg.Version = value;
            break;
        case 'p': // Provides
            for (var _i = 0, _a = value.split(' '); _i < _a.length; _i++) {
                var name = _a[_i];
                name = name.split('=')[0];
                curPkg.Provides.push(name);
            }
            break;
        case 'r': // Depends
        case 'D': // Depends
            // tslint:disable-next-line:no-duplicate-variable
            for (var _b = 0, _c = value.split(' '); _b < _c.length; _b++) {
                var name = _c[_b];
                if (name.charAt(0) !== '!') {
                    name = name.split('=')[0];
                    curPkg.Deps[name] = true;
                }
            }
            break;
    }
    return curPkg;
}
